File: weather-app/src/main/java/rcgonzalezf/org/weather/location/LocationManager.java
package rcgonzalezf.org.weather.location;

import android.Manifest;
import android.os.Build;
import android.support.annotation.NonNull;
import android.support.annotation.VisibleForTesting;
import android.support.design.widget.Snackbar;
import android.view.View;
import java.lang.ref.WeakReference;
import rcgonzalezf.org.weather.R;
import rcgonzalezf.org.weather.common.BaseActivity;
import rcgonzalezf.org.weather.common.PermissionChecker;
import rcgonzalezf.org.weather.common.PermissionResultListener;

public class LocationManager implements LocationRetrieverListener {

  @VisibleForTesting static int CURRENT_SDK_INT = Build.VERSION.SDK_INT;
  private final WeakReference<BaseActivity> baseActivityWeakReference;
  private final WeakReference<View> mContentWeakReference;
  private final LocationRetriever mLocationRetriever;
  @VisibleForTesting PermissionChecker mPermissionChecker;

  public LocationManager(@NonNull BaseActivity baseActivity, View content) {
    this.baseActivityWeakReference = new WeakReference<>(baseActivity);
    this.mContentWeakReference = new WeakReference<>(content);
    this.mLocationRetriever = new LocationRetriever(baseActivity, this);
  }

  @Override public void checkForPermissions() {
    BaseActivity baseActivity = baseActivityWeakReference.get();
    View content = mContentWeakReference.get();
    checkForPermissions(baseActivity, content);
  }

  @Override public void onEmptyLocation() {
    BaseActivity baseActivity = baseActivityWeakReference.get();
    View content = mContentWeakReference.get();
    onEmptyLocation(baseActivity, content);
  }

  @Override public void onLocationFound(double lat, double lon) {
    BaseActivity baseActivity = baseActivityWeakReference.get();
    onLocationFound(baseActivity, lat, lon);
  }

  public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
      @NonNull int[] grantResults) {
    if (mPermissionChecker != null && CURRENT_SDK_INT >= Build.VERSION_CODES.M) {
      //noinspection NewApi
      mPermissionChecker.onRequestPermissionsResult(requestCode, permissions, grantResults);
    }
  }

  public void connect() {
    mLocationRetriever.connect();
  }

  public void disconnect() {
    mLocationRetriever.disconnect();
  }

  @VisibleForTesting @NonNull PermissionResultListener getPermissionResultListener() {
    return new PermissionResultListener() {

      @Override public void onSuccess() {
        mLocationRetriever.onLocationPermissionsGranted();
      }

      @Override public void onFailure() {
        mLocationRetriever.onLocationPermissionFailure();
      }
    };
  }

  @VisibleForTesting void checkForPermissions(BaseActivity baseActivity, View content) {
    if (baseActivity != null && content != null) {
      mPermissionChecker =
          new PermissionChecker(Manifest.permission.ACCESS_FINE_LOCATION, baseActivity,
              PermissionChecker.LOCATION, content, R.string.permissions_location_granted,
              R.string.permissions_location_not_granted, R.string.permissions_location_rationale);

      if (mPermissionChecker.hasPermission()) {
        mLocationRetriever.onLocationPermissionsGranted();
      } else {
        mPermissionChecker.requestPermission(getPermissionResultListener());
      }
    }
  }

  @VisibleForTesting void onEmptyLocation(final BaseActivity baseActivity, View content) {
    if (baseActivity != null && content != null) {
      Snackbar.make(content, baseActivity.getString(R.string.location_off_msg),
          Snackbar.LENGTH_SHORT).show();
    }
  }

  @VisibleForTesting void onLocationFound(BaseActivity baseActivity, double lat, double lon) {
    if (baseActivity != null) {
      baseActivity.searchByLocation(lat, lon);
    }
  }
}

File: weather-app/src/main/java/rcgonzalezf/org/weather/location/LocationManager.kt
package rcgonzalezf.org.weather.location

import android.Manifest.permission.ACCESS_FINE_LOCATION
import android.os.Build
import android.view.View
import androidx.annotation.RequiresApi
import androidx.annotation.VisibleForTesting
import com.google.android.material.snackbar.Snackbar
import rcgonzalezf.org.weather.R
import rcgonzalezf.org.weather.common.BaseActivity
import rcgonzalezf.org.weather.common.PermissionChecker
import rcgonzalezf.org.weather.common.PermissionChecker.Companion.LOCATION
import rcgonzalezf.org.weather.common.PermissionResultListener
import java.lang.ref.WeakReference

open // for Mockito
class LocationManager(baseActivity: BaseActivity,
                      private val locationSearch: LocationSearch, content: View)
    : LocationRetrieverListener {
    private val baseActivityWeakReference: WeakReference<BaseActivity> = WeakReference(baseActivity)
    private val contentWeakReference: WeakReference<View> = WeakReference(content)
    private val locationRetriever: LocationRetriever = LocationRetriever(baseActivity, this)

    companion object {
        @JvmField
        @VisibleForTesting
        var CURRENT_SDK_INT = Build.VERSION.SDK_INT
    }

    @JvmField
    @VisibleForTesting
    var permissionChecker: PermissionChecker? = null

    override fun checkForPermissions() {
        val baseActivity = baseActivityWeakReference.get()
        val content = contentWeakReference.get()
        checkForPermissions(baseActivity, content)
    }

    override fun onEmptyLocation() {
        val baseActivity = baseActivityWeakReference.get()
        val content = contentWeakReference.get()
        onEmptyLocation(baseActivity, content)
    }

    override fun onLocationFound(lat: Double, lon: Double) {
        locationSearch.searchByLatLon(lat, lon)
    }

    @RequiresApi(Build.VERSION_CODES.M)
    fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>,
                                   grantResults: IntArray) {
        if (CURRENT_SDK_INT >= Build.VERSION_CODES.M) {
            permissionChecker?.onRequestPermissionsResult(requestCode, permissions, grantResults)
        }
    }

    open fun connect() {
        locationRetriever.connect()
    }

    open fun disconnect() {
        locationRetriever.disconnect()
    }

    @get:VisibleForTesting
    val permissionResultListener: PermissionResultListener
        get() = object : PermissionResultListener {
            override fun onSuccess() {
                locationRetriever.onLocationPermissionsGranted()
            }

            override fun onFailure() {
                locationRetriever.onLocationPermissionFailure()
            }
        }

    @VisibleForTesting
    fun checkForPermissions(baseActivity: BaseActivity?, content: View?) {
        if (baseActivity != null && content != null) {
            permissionChecker = PermissionChecker(ACCESS_FINE_LOCATION, baseActivity,
                    LOCATION, content, R.string.permissions_location_granted,
                    R.string.permissions_location_not_granted,
                    R.string.permissions_location_rationale)
            permissionChecker?.let {
                if (it.hasPermission()) {
                    locationRetriever.onLocationPermissionsGranted()
                } else {
                    it.requestPermission(permissionResultListener)
                }
            }
        }
    }

    @VisibleForTesting
    fun onEmptyLocation(baseActivity: BaseActivity?, content: View?) {
        if (baseActivity != null && content != null) {
            Snackbar.make(content, baseActivity.getString(R.string.location_off_msg),
                    Snackbar.LENGTH_SHORT).show()
        }
    }
}
